subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/VersionControlConsumerConnection.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.internal.consumer.connection;
+
+import org.gradle.api.UncheckedIOException;
+import org.gradle.tooling.BuildAction;
+import org.gradle.tooling.UnsupportedVersionException;
+import org.gradle.tooling.internal.consumer.TestExecutionRequest;
+import org.gradle.tooling.internal.consumer.parameters.ConsumerOperationParameters;
+import org.gradle.tooling.internal.consumer.versioning.VersionDetails;
+import org.gradle.tooling.internal.protocol.ConnectionVersion4;
+import org.gradle.tooling.model.internal.Exceptions;
+import org.gradle.util.GradleVersion;
+
+import java.io.IOException;
+
+public class VersionControlConsumerConnection implements ConsumerConnection {
```

Comment:
❌ This class is more complicated than it needs to be because it handles too many cases:

- deprecated
- unsupported
- supported

Instead, we should have a `DeprecatedConsumerConnection` which we use only when the version is deprecated. For the unsupported case we should keep the `UnsupportedOlderVersionConnection `. And for the fully supported case, we should do no wrapping at all, as it would just add unnecessary work.

You'll see that this removes the need for if-conditions, making the code shorter and clearer.